profilers/attachments/perfTemplate.c

```diff
@@ -29,48 +33,55 @@ int set_up_perf_event(struct perf_event_attr* pe, unsigned long long config) {
     fd = perf_event_open(pe, 0, -1, -1, 0);
     if (fd == -1) {
         fprintf(stderr, "Error opening leader %llx\n", pe->config);
-        exit(EXIT_FAILURE);
+        // exit(EXIT_FAILURE);
     }
+    free(pe);
     return fd;
 }
 
-unsigned long long* traceFunction(traced_fun_t traced, unsigned long long* configs, int configs_len) {
-    struct perf_event_attr pe[configs_len];
-    unsigned long long* results = calloc(configs_len, sizeof(*results));
-    int fd[configs_len];
+static void init() {
+    size_t configs_len = sizeof(configs) / sizeof(*configs);
+    perf_fd_len = configs_len;
+    perf_fd = calloc(configs_len, sizeof(*perf_fd));
     for (int i = 0; i < configs_len; i++) {
-        fd[i] = set_up_perf_event(&pe[i], configs[i]);
+        perf_fd[i] = set_up_perf_event(configs[i]);
     }
 
     for (int i = 0; i < configs_len; i++) {
-        ioctl(fd[i], PERF_EVENT_IOC_RESET, 0);
+        if (perf_fd[i] != -1)
+            ioctl(perf_fd[i], PERF_EVENT_IOC_RESET, 0);
     }
 
     for (int i = 0; i < configs_len; i++) {
-        ioctl(fd[i], PERF_EVENT_IOC_ENABLE, 0);
+        if (perf_fd[i] != -1)
+            ioctl(perf_fd[i], PERF_EVENT_IOC_ENABLE, 0);
     }
+}
 
-    traced();
-
-    for (int i = 0; i < configs_len; i++) {
-        ioctl(fd[i], PERF_EVENT_IOC_DISABLE, 0);
-    }
+static void fin() {
```

Comment:
I recommend splitting it to 2 functions:
- Signal handler
- Function that collecs perf data, prints result and closes fds